src/Microsoft.AspNet.Mvc.Core/ActionConstraints/ConsumesAttribute.cs

```diff
@@ -0,0 +1,135 @@
+        {
+            MediaTypeHeaderValue requestContentType = null;
+            MediaTypeHeaderValue.TryParse(context.HttpContext.Request.ContentType, out requestContentType);
+            if (context.ActionDescriptor.FilterDescriptors.Last().Filter == this)
+            {
+                // Only execute if this is the last filter before calling the action.
+                // This is because we want to avoid running filters which were inherited when there was already a filter
+                // on the action. This ensures that we only run the filter which is closest to the action.
+                if (requestContentType != null &&
+                    !ContentTypes.Any(contentType => contentType.IsSubsetOf(requestContentType)))
+                {
+                    context.Result = new UnsupportedMediaTypeResult();
+                }
+            }
+        }
+
+        public void OnResourceExecuted([NotNull] ResourceExecutedContext context)
+        {
+        }
+
+        private bool IsApplicable(ActionConstraintContext context)
+        {
+            return context.CurrentCandidate.Action.FilterDescriptors.Last(filter => filter.Filter is ConsumesAttribute).Filter == this;
+        }
+
+        public bool Accept(ActionConstraintContext context)
+        {
+            // First check to see if this is applicble
+            if (!IsApplicable(context))
+            {
+                // since this constraint is not applicable it should return true.
+                return true;
+            }
+
+            MediaTypeHeaderValue requestContentType = null;
+            MediaTypeHeaderValue.TryParse(context.RouteContext.HttpContext.Request.ContentType, out requestContentType);
+
+            // If the request content type is null we need to act like pass through.
+            // In case there is a single candidate with a constraint it should be selected.
+            // If there are multiple actions with consumes action constraints this should result in ambiguous exception
+            // unless there is another action without a consumes constraint.
+            if (requestContentType == null)
+            {
+                var isActionWithoutConsumeConstraintPresent = context.Candidates.Any(
+                    candidate => candidate.Constraints == null ||
+                    !candidate.Constraints.Any(constraint => constraint is IConsumesActionConstraint));
+
+                return !isActionWithoutConsumeConstraintPresent;
+            }
+
+            if (ContentTypes.Any(c => c.IsSubsetOf(requestContentType)))
+            {
+                return true;
+            }
+
+            var firstCandidate = context.Candidates[0];
+            if (firstCandidate != context.CurrentCandidate)
+            {
+                // If the current candidate has reached here, its not a match.
+                return false;
+            }
+
+            foreach (var candidate in context.Candidates)
```

Comment:
It would be good to expand on this with a comment as well for future generations. Something like:

```
Run the matching logic for all `IConsumesActionConstraints` we can find, and see what matches.
1). If we have a unique best match, then only that constraint should return true
2). If we have multiple matches, then all constraints that match will return true, resulting in ambiguity (maybe)
3). If we have no matches, then choose an action (first or last, your pick) to return true. It will later return a 415
```